### PR TITLE
 Fuzztests for User Profile service namedpipe pipe

### PR DESCRIPTION
**Pull request by karthiknadig**
1. Added tests for the user profile service named pipe.
2. Added timed cancellations to service pipe.

### CODE REVIEW

File: src/Host/Protocol/Test/UserProfileServicePipe/UserProfileServicePipeTest.cs [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+        // empty username domain password
+        [InlineData("{\"Username\": \"\", \"Domain\": \"\", \"Password\": \"\"}", "", "", "", true, false, false)]
+        // whitespace input string
+        [InlineData("                     ", null, null, null, true, false, false)]
+        public async Task CreateProfileTest(string input, string username, string domain, string password, bool isValidParse, bool isValidAccount, bool isExistingAccount) {
+            var creator = UserProfileCreatorMock.Create(username, domain, password, isValidParse, isValidAccount, isExistingAccount);
+            await CreateProfileTestRunnerAsync(creator, input, isValidParse, isValidAccount, isExistingAccount);
+        }
+
+        [Test]
```
Comment (int19h) [inline src/Host/Protocol/Test/UserProfileServicePipe/UserProfileServicePipeTest.cs]: We should probably have a distinct category for fuzz tests. Take a look at how other test classes use test category attributes.

File: src/Host/Protocol/Test/UserProfileServicePipe/UserProfileServicePipeTest.cs [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+        public async Task CreateProfileFuzzTest() {
+            string inner = "\"Username\": {0}, \"Domain\": {1}, \"Password\":{2}";
+            List<string> testResults = new List<string>();
```
Comment (int19h) [inline src/Host/Protocol/Test/UserProfileServicePipe/UserProfileServicePipeTest.cs]: Unused.

File: src/Host/Protocol/Test/UserProfileServicePipe/UserProfileServicePipeTest.cs [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+                    await CreateProfileTestRunnerAsync(creator, json, false, false, false, true);
+                } catch (IOException) {
+                    // expect pipe to fail. The client side pipe throws an IOException when the server side pipe 
+                    // closes due to IO error or attempt to access unauthorized memory.
+                }
+            }
+        }
+
+        private byte[] GenerateBytes() {
+            Random rd = new Random((int)DateTime.Now.Ticks);
```
Comment (int19h) [inline src/Host/Protocol/Test/UserProfileServicePipe/UserProfileServicePipeTest.cs]: You can just use the default constructor for `Random` - it automatically uses ticks or something similar.

File: src/Host/Protocol/Test/UserProfileServicePipe/UserProfileServicePipeTest.cs [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+        // Missing closing paren
+        [InlineData("{", null, null, null, false, false, false)]
+        // empty json object
+        [InlineData("{}", null, null, null, false, false, false)]
+        // No username, domain and password
+        [InlineData("{\"Username\":, \"Domain\":, \"Password\":}", null, null, null, false, false, false)]
+        // empty username domain password
+        [InlineData("{\"Username\": \"\", \"Domain\": \"\", \"Password\": \"\"}", "", "", "", true, false, false)]
+        // whitespace input string
+        [InlineData("                     ", null, null, null, true, false, false)]
```
Comment (int19h) [inline src/Host/Protocol/Test/UserProfileServicePipe/UserProfileServicePipeTest.cs]: Can you also add an explicit test case for zero-length input?

**Comment by karthiknadig** [comment]
Ping

File: src/Common/Core/Impl/OS/IUserCredentials.cs [hunk 5]
```diff
@@ -0,0 +1,10 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+namespace Microsoft.Common.Core.OS {
+    public interface IUserCredentials {
+        string Username { get; set; }
+        string Password { get; set; }
```
Comment (MikhailArkhipov) [inline src/Common/Core/Impl/OS/IUserCredentials.cs]: Scares me a bit. SecureString?

File: src/Host/Protocol/Impl/RUserProfileCreatorImpl.cs [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+                } else {
+                    logger?.LogError(Resources.Error_UserProfileDirectoryWasNotFound, credentails.Domain, credentails.Username, Marshal.GetLastWin32Error());
+                    result = RUserProfileCreateResponse.Create((uint)Marshal.GetLastWin32Error(), profileExists, profileDir.ToString());
+                }
+            } else {
+                logger?.LogError(Resources.Error_UserLogonFailed, credentails.Domain, credentails.Username, Marshal.GetLastWin32Error());
+                result = RUserProfileCreateResponse.Create((uint)Marshal.GetLastWin32Error(), false, null);
+            }
+
+            if (token != IntPtr.Zero) {
```
Comment (MikhailArkhipov) [inline src/Host/Protocol/Impl/RUserProfileCreatorImpl.cs]: Probably should be in try-finally

File: src/Host/Protocol/Test/UserProfileServicePipe/UserProfileServicePipeTest.cs [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+                    if (isValidParse) {
+                        result.Error.Should().Be((uint)(isValidAccount ? 0 : 13));
+                        result.ProfileExists.Should().Be(isExistingAccount);
+                    } else {
+                        result.Should().BeNull();
+                    }
+                }
+            }
+
+            testDone.Wait();
```
Comment (MikhailArkhipov) [inline src/Host/Protocol/Test/UserProfileServicePipe/UserProfileServicePipeTest.cs]: We have async tests, they use regular awaits and IAsyncLifetime. Otherwise if test hangs, it may block all subsequent tests

File: src/Host/Protocol/Test/UserProfileServicePipe/UserProfileServicePipeTest.cs [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+                byte[] responseRaw = new byte[1024];
+                var bytesRead = await client.ReadAsync(responseRaw, 0, responseRaw.Length, ct);
+                jsonResp = Encoding.Unicode.GetString(responseRaw, 0, bytesRead);
+            }
+            return JsonConvert.DeserializeObject<UserProfileResultMock>(jsonResp);
+        }
+
+        private async Task CreateProfileTestRunnerAsync(IUserProfileServices creator, string input, bool isValidParse, bool isValidAccount, bool isExistingAccount, int serverTimeOut, int clientTimeOut, bool isFuzzTest = false) {
+            ManualResetEventSlim testDone = new ManualResetEventSlim(false);
+            Task.Run(async () => {
```
Comment (MikhailArkhipov) [inline src/Host/Protocol/Test/UserProfileServicePipe/UserProfileServicePipeTest.cs]: I'd split method into fuzz part and regular part

File: src/Host/Protocol/Test/UserProfileServicePipe/UserProfileServicePipeTest.cs [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+                            f.ShouldNotThrow();
+                        } else {
+                            Func<Task> f = () => RUserProfileCreator.CreateProfileAsync(serverTimeOutms: serverTimeOut, clientTimeOutms: clientTimeOut, userProfileService: creator);
+                            await f.ShouldThrowAsync<Exception>();
+                        }
+                    }
+                } finally {
+                    testDone.Set();
+                }
+            }).DoNotWait();
```
Comment (MikhailArkhipov) [inline src/Host/Protocol/Test/UserProfileServicePipe/UserProfileServicePipeTest.cs]: Why we can't await it?

File: src/Host/Protocol/Test/Microsoft.R.Host.Protocol.Test.csproj [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+  <ItemGroup>
+  </ItemGroup>
+  <ItemGroup>
+    <FilesToSign Include="$(OutputPath)\$(AssemblyName).dll">
+      <Authenticode>Microsoft</Authenticode>
+      <StrongName>StrongName</StrongName>
+    </FilesToSign>
+  </ItemGroup>
+  <Import Project="$(VSToolsPath)\TeamTest\Microsoft.TestTools.targets" Condition="Exists('$(VSToolsPath)\TeamTest\Microsoft.TestTools.targets')" />
+  <Import Project="$(MSBuildToolsPath)\Microsoft.CSharp.targets" />
```
Comment (MikhailArkhipov) [inline src/Host/Protocol/Test/Microsoft.R.Host.Protocol.Test.csproj]: MicroBuild targets and props

File: src/Host/Protocol/Test/UserProfileServicePipe/UserProfileServicePipeTest.cs [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+        [InlineData("                     ", null, null, null, true, false, false)]
+        // empty string
+        [InlineData("", null, null, null, false, false, false)]
+        public async Task CreateProfileTest(string input, string username, string domain, string password, bool isValidParse, bool isValidAccount, bool isExistingAccount) {
+            var creator = UserProfileCreatorMock.Create(username, domain, password, isValidParse, isValidAccount, isExistingAccount);
+            await CreateProfileTestRunnerAsync(creator, input, isValidParse, isValidAccount, isExistingAccount, 500, 500);
+        }
+
+        [Test]
+        [Category.FuzzTest]
```
Comment (int19h) [inline src/Host/Protocol/Test/UserProfileServicePipe/UserProfileServicePipeTest.cs]: I suggested having a generic category for fuzz stuff, so that they could be excluded from test runs altogether (to get deterministic runs if we need to).

File: src/Host/Protocol/Test/UserProfileServicePipe/UserProfileServicePipeTest.cs (hunk 7 above)
Comment (karthiknadig) [inline src/Host/Protocol/Test/UserProfileServicePipe/UserProfileServicePipeTest.cs]: See explanation below for this. we can put a time out here. Note, if this test ends without the server disconnecting, the next test will certainly fail. This wait is needed to ensure all threads are done.

File: src/Host/Protocol/Test/UserProfileServicePipe/UserProfileServicePipeTest.cs (hunk 4 above)
Comment (karthiknadig) [inline src/Host/Protocol/Test/UserProfileServicePipe/UserProfileServicePipeTest.cs]: added

File: src/Host/Protocol/Test/Microsoft.R.Host.Protocol.Test.csproj (hunk 10 above)
Comment (karthiknadig) [inline src/Host/Protocol/Test/Microsoft.R.Host.Protocol.Test.csproj]: I don't see that in other test projects. Should it be there?

File: src/Host/Protocol/Impl/RUserProfileCreator.cs [hunk 12]
```diff
@@ -83,7 +83,7 @@ public class RUserProfileCreator {
 
             string json = Encoding.Unicode.GetString(requestRaw, 0, bytesRead);
 
-            var requestData = JsonConvert.DeserializeObject<RUserProfileCreateRequest>(json);
+            var requestData = JsonConvert.DeserializeObject<RUserProfileCreateRequest>(json, new SecureStringJsonConverter());
```
Comment (int19h) [inline src/Host/Protocol/Impl/RUserProfileCreator.cs]: You can also stick `JsonConverterAttribute` onto the property in the class. Then you won't have to repeat the converter on every call to `Serialize` or `Deserialize`.

File: src/Host/Protocol/Test/UserProfileServicePipe/UserProfileServicePipeTest.cs (hunk 9 above)
Comment (MikhailArkhipov) [inline src/Host/Protocol/Test/UserProfileServicePipe/UserProfileServicePipeTest.cs]: OK

File: src/Host/Protocol/Test/Microsoft.R.Host.Protocol.Test.csproj (hunk 10 above)
Comment (MikhailArkhipov) [inline src/Host/Protocol/Test/Microsoft.R.Host.Protocol.Test.csproj]: Hm interesting. OK, `"MicroBuild.Core": "0.2.0",` in project.json I guess would be sufficient.

**Comment by int19h** [comment]
👍
